File: Java/090__Subsets_II.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        if (nums == null || nums.length == 0)
            return Collections.emptyList();
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        subSetHelper(nums, new ArrayList<Integer>(), result, 0);
        return result;
    }

    private void subSetHelper(int[] nums, List<Integer> tempList, List<List<Integer>> result, int start) {
        result.add(new ArrayList<Integer>(tempList));
        for (int i = start; i < nums.length; i++) {
            if (i > start && nums[i] == nums[i - 1])
                continue;
            tempList.add(nums[i]);
            subSetHelper(nums, tempList, result, i + 1);
            tempList.remove(tempList.size() - 1);
        }
    }
}

File: Python/090__Subsets_II.py
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = [[]]
        nums.sort()
        for val in nums:
            result += [lst + [val] for lst in result if lst + [val] not in result]
        return result


# Another Solution
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if not nums:
            return []

        nums.sort()
        res, cur = [[]], []

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                cur = [item + [nums[i]] for item in cur]
            else:
                cur = [item + [nums[i]] for item in res]
            res += cur
        return res
